pkg/controller/framework/controller.go

```diff
@@ -94,7 +98,38 @@ func (c *Controller) Run(stopCh <-chan struct{}) {
 
 	r.RunUntil(stopCh)
 
+	// Spawn a number of workers.
+	// TODO: Make it an argument.
+	c.workers = 16
+	wg := sync.WaitGroup{}
+	wg.Add(c.workers)
+	c.workQueue = make([]chan interface{}, c.workers)
+	for i := 0; i < c.workers; i++ {
+		// TODO: What should be the size of channel?
+		c.workQueue[i] = make(chan interface{}, 10)
+		go func(worker int) {
+			// TODO: Handle crash
+			defer wg.Done()
+			for {
+				obj, ok := <-c.workQueue[worker]
+				if !ok {
+					return
+				}
+				err := c.config.Process(obj)
+				if err != nil {
+					if c.config.RetryOnError {
+						// This is the safe way to re-enqueue.
+						c.config.Queue.AddIfNotPresent(obj)
```

Comment:
One challenge here is that it can result in very fast retries of conditions that don't immediately clear up.  We put a rate limiter on retries to slow the work queue (burst 10, qps 2-3/s) in some of our controllers.